doc/manual/documentation.rst

```diff
@@ -18,25 +24,113 @@ In 0.4:
 
 Documentation is interpreted as `Markdown <https://en.wikipedia.org/wiki/Markdown>`_,
 so you can use indentation and code fences to delimit code examples from text.
+Technically, any object can be associated with any other as metadata;
+Markdown happens to be the default, but one can construct other string
+macros and pass them to the ``@doc`` macro just as well.
+
+Here is a more complex example, still using Markdown:
 
 .. doctest::
 
     """
-    The `@bar` macro will probably make your code 2x faster or something. Use
-    it like this:
+        bar(x[, y])
 
-        @bar buy_drink_for("Jiahao")
+    Computes the Bar index between `x` and `y`. If `y` is missing, the function computes
+    the index between all pairs of columns of `x`.
+
+    # Examples
+    ```julia
+    julia> bar([1, 2], [1, 2])
+    1
+    ```
     """
-    macro bar(ex) ...
+    function bar(x, y) ...
 
-Documentation is very free-form; there are no set formatting
-restrictions or strict conventions. It's hoped that best practices will
-emerge fairly naturally as package developers learn to use the new
-system.
+Documentation is very free-form, though we recommend following some simple
+conventions, illustrated in the above example.
 
-Technically, any object can be associated with any other as metadata;
-Markdown happens to be the default, but one can construct other string
-macros and pass them to the ``@doc`` macro just as well.
+1. Always show the signature of a function at the top of the documentation,
+with a four-space indent so that it is printed as Julia code.
+
+  This can be identical to the signature present in the Julia code
+  (like ``mean(x::AbstractArray)``), or a simplified form.
+  Optional arguments should be put into brackets ``[]``, with their default values
+  when applicable. An alternative solution is to use several lines: one without
+  optional arguments, the other(s) with them. This solution can also be used to
+  document several related methods of a given function.
```

Comment:
I personally think using real julia syntax for this would be clearer than square brackets. There's precedent for brackets with optional args, but it's not real syntax.